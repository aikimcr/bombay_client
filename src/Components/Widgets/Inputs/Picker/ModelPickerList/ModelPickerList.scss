@use "@styles/mixins.scss";

.model-picker-list {
  --picker-color: var(--primary-button-color);
  --picker-background-color: var(--color-white-245);
  --picker-font-size: var(--standard-button-font-size);

  margin: 0;
  padding: 0;
  border: none;

  height: 100%;
  width: 100%;
  max-height: 25cap;
  max-width: 25ch;
  overflow: hidden;
  @include mixins.scrollListContainer;
  position-anchor: --picker-button;
  position-area: y-end span-right;
  position-try: normal flip-block;

  &__list {
    --input-background-color: var(--picker-color);

    display: grid;
    grid-auto-flow: row;
    gap: 0.5em;
    padding: 0.5em;
    margin: 0 0 0 .5rem;
    justify-content: start;
    align-content: center;
    color: var(--picker-color);
    background-color: var(--picker-background-color);

    @include mixins.inputBorder($borderRadius: 1rem, $borderStyle: solid, $borderWidth: 1px);

    font-size: var(--standard-button-font-size);
    line-height: 1;
    padding: 0.25em 0.5em;

    li {
      cursor: pointer;
      color: var(--picker-color);
      background-color: var(--picker-background-color);
      width: 100%;

      &:hover {
        background-color: rgb(80%, 80%, 80%);
      }

      &:active {
        color: rgb(10%, 10%, 10%);
        background-color: rgb(90%, 90%, 90%);
      }
    }

    li::marker {
      content: '';
      font-size: 1px;
    }
  }
}
