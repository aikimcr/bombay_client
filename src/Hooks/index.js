export * from './useLoginTracking';
export * from './useModelCollection';
