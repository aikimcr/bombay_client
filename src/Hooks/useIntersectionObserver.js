import { useEffect, useRef } from "react";

export function useIntersectionObserver(topRef, callback, options = {}) {
  const observer = useRef(null);

  useEffect(() => {
    if (topRef.current == null) return;

    if (observer.current == null) {
      observer.current = new IntersectionObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      observer.current.disconnect();
    };
  }, [callback, options, topRef]);

  return observer;
}

export default useIntersectionObserver;
