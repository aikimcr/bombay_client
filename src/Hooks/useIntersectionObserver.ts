import { useEffect, useRef, RefObject } from 'react';

export interface IntersectionObserverOptions {
  root?: Element | null;
  rootMargin?: string;
  threshold?: number | number[];
}

export interface UseIntersectionObserverReturn {
  current: IntersectionObserver | null;
}

export function useIntersectionObserver(
  topRef: RefObject<Element>,
  callback: IntersectionObserverCallback,
  options: IntersectionObserverOptions = {},
): UseIntersectionObserverReturn {
  const observer = useRef<IntersectionObserver | null>(null);

  useEffect(() => {
    if (topRef.current == null) {
      return;
    }

    if (observer.current == null) {
      observer.current = new IntersectionObserver(callback, {
        ...options,
        root: topRef.current.parentElement,
      });
    }

    return () => {
      if (observer.current) {
        observer.current.disconnect();
      }
    };
  }, [callback, options, topRef]);

  return observer;
}

export default useIntersectionObserver;
