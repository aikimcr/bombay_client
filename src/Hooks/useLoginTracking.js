import { useCallback, useEffect, useRef, useState } from "react";

import { loginStatus, refreshToken } from "../Network/Login";
// Allow external forcing of the state
let setLoggedIn;
export function forceLoginState(newState) {
  if (setLoggedIn) return setLoggedIn(newState);
}

function toMilliseconds(minutes) {
  return minutes * 60 * 1000;
}

export function useLoginTracking() {
  const [loginState, setLoginState] = useState(false);
  const [activitySeen, setActivitySeen] = useState(true);

  const activityCheckMinutes = 30;
  const loginCheckMinutes = 10;

  let activityHandle = useRef(false);

  setLoggedIn = setLoginState; // Allow external forcing of the state.

  const checkLoginState = useCallback(async function () {
    const result = await loginStatus(activityCheckMinutes);
    setLoginState(result);
    return result;
  }, []);

  const refreshState = useCallback(
    async function () {
      if (!loginState) return false;

      const result = await refreshToken();
      setLoginState(result != null);
      return result != null;
    },
    [loginState],
  );

  const checkActivity = useCallback(function (evt) {
    // console.debug(`${new Date().toLocaleString()}: Check Activity (${evt}) (Activity: ${activitySeen}, loginState: ${loginState})`);

    clearTimeout(activityHandle.current);

    if (evt != null) setActivitySeen(true);

    activityHandle.current = setTimeout(async () => {
      // console.debug(`${new Date().toLocaleString()}: Clear Activity (Activity: ${activitySeen}, loginState: ${loginState})`);

      setActivitySeen(false);
    }, toMilliseconds(activityCheckMinutes));
  }, []);

  const setListeners = useCallback(() => {
    document.addEventListener("keydown", checkActivity);
    document.addEventListener("mousedown", checkActivity);
    document.addEventListener("mousemove", checkActivity);
  }, [checkActivity]);

  const removeListeners = useCallback(() => {
    document.removeEventListener("mousemove", checkActivity);
    document.removeEventListener("mousedown", checkActivity);
    document.removeEventListener("keydown", checkActivity);
  }, [checkActivity]);

  const checkLogin = useCallback(
    async function () {
      // console.debug(`${new Date().toLocaleString()}: Check login (Activity: ${activitySeen}, loginState: ${loginState})`);

      if (activitySeen) {
        await refreshState();
        return;
      }

      return await checkLoginState();
    },
    [activitySeen, refreshState, checkLoginState],
  );

  useEffect(() => {
    // console.debug(`${new Date().toLocaleString()}: Change (Activity: ${activitySeen}, loginState: ${loginState})`);
    let intervalHandle = false;

    if (loginState) {
      intervalHandle = setInterval(
        checkLogin,
        toMilliseconds(loginCheckMinutes),
      );
      checkActivity();
      setListeners();
    }

    return () => {
      removeListeners();
      clearTimeout(activityHandle.current);
      clearInterval(intervalHandle);
    };
  }, [
    activitySeen,
    loginState,
    activityHandle,
    checkActivity,
    checkLogin,
    removeListeners,
    setListeners,
  ]);

  useEffect(() => {
    checkLoginState();
  }, [checkLoginState]);

  return [loginState, setLoginState];
}

export default useLoginTracking;
