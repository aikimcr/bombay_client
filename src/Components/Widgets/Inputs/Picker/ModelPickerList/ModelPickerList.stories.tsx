import React from 'react';
import type { Meta, StoryObj } from '@storybook/react-webpack5';

import {
  ContextChanger,
  TestCollectionHandlers,
  TestCollectionOne,
} from '../../../../../testHelpers';

import { ModelPickerList } from './ModelPickerList';

const meta = {
  component: ModelPickerList,

  tags: ['autodocs'],
} satisfies Meta<typeof ModelPickerList>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  parameters: {
    msw: {
      handlers: TestCollectionHandlers,
    },
  },
  args: {
    initialCollection: new TestCollectionOne({}),
  },
  render: (args) => (
    <ContextChanger initialLoggedIn={true}>
      <ModelPickerList {...args} isOpen={true} />
    </ContextChanger>
  ),
};
