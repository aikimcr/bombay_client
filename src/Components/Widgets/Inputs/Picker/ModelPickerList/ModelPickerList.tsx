import React, { useEffect, useRef, useState } from 'react';
import { useModelCollection } from '../../../../../Hooks';
import { CollectionBase } from '../../../../../Model/CollectionBase';
import { ModelBase } from '../../../../../Model/ModelBase';

import './ModelPickerList.scss';

interface ModelPickerListProps {
  id?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  initialCollection: CollectionBase<any, any>;
  isOpen?: boolean; // for testing.
}

export const ModelPickerList: React.FC<ModelPickerListProps> = ({
  id = 'model-picker-list',
  initialCollection,
  isOpen = false,
}): React.ReactElement => {
  const listRef = useRef<HTMLUListElement>(null);

  const [isMounted, setIsMounted] = useState(false);
  useModelCollection({
    initialCollection,
    topRef: listRef,
    isMounted,
  });

  useEffect(() => {
    setIsMounted(true);
  }, []);

  return (
    <dialog
      id={id}
      className="model-picker-list"
      data-testid="model-picker-list"
      popover="auto"
      open={isOpen}
    >
      <ul
        ref={listRef}
        className="model-picker-list__list"
        data-testid="model-picker-list-list"
      >
        {initialCollection.map((model: ModelBase) => {
          const key = model.id;
          return (
            <li key={key} data-modelid={model.id}>
              {model.name}
            </li>
          );
        })}
      </ul>
    </dialog>
  );
};
