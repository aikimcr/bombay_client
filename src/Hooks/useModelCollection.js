import { useState, useEffect, useContext, useRef, useCallback } from 'react';

import BombayLoginContext from '../Context/BombayLoginContext';
import useIntersectionObserver from './useIntersectionObserver';

export const useModelCollection = ({ CollectionClass, topRef }) => {
  const { loggedIn } = useContext(BombayLoginContext);
  const [shouldPage, setShouldPage] = useState(false);
  const [loading, setLoading] = useState(loggedIn);

  const [collection, setCollection] = useState(null);
  const observer = useIntersectionObserver(topRef, (entries, observer) => {
    entries.forEach((entry) => {
      if (entry.isIntersecting) {
        setShouldPage(true);
        observer.unobserve(entry.target);
      }

      // I haven't found a case where this is actually needed, but I haven't ruled it out.
      // if (entry.isVisible) {
      // }
    });
  });

  const refreshCollection = async () => {
    // console.count(`refresh, ${loggedIn}`);
    if (!loggedIn) {
      setCollection(null);
      return;
    }

    setLoading(true);

    const newCollection = new CollectionClass();
    await newCollection.ready();
    // console.count(`newCollection ready: ${newCollection}`);

    setLoading(false);
    setCollection(newCollection);
  };

  useEffect(() => {
    // console.count(`mount hook, ${collection}, ${loading}`);
    if (collection == null) {
      refreshCollection();
    }
  }, []);

  useEffect(() => {
    // console.count(`shouldPage: ${shouldPage}, loading: ${loading}`);
    if (!shouldPage) {
      return;
    }
    if (loading) {
      return;
    }

    setShouldPage(false);

    if (collection.hasNextPage()) {
      setLoading(true);

      collection.fetchNextPage().then(() => {
        setLoading(false);
      });
    }
  }, [shouldPage, loading]);

  useEffect(() => {
    // console.count(`loading: ${loading}, observer: ${observer}, topRef: ${topRef}`);
    if (loading) {
      return;
    }

    if (collection && collection.hasNextPage()) {
      const myElement = topRef.current?.querySelector('li:last-child');

      if (myElement) {
        observer.current.observe(myElement);
      }
    }
  }, [loading, observer, topRef]);

  useEffect(() => {
    // console.count(`loggedIn change, ${loggedIn}, ${collection} ${loading}`);
    if (loading) {
      return;
    }

    if (loggedIn && collection == null) {
      refreshCollection();
      return;
    }

    if (!loggedIn) {
      setCollection(null);
    }
  }, [loggedIn]);

  // console.log('return collection', collection);
  return [collection, refreshCollection];
};

export default useModelCollection;
