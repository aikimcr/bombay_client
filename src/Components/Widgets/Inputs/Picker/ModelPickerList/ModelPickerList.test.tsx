import React from 'react';

import { act, render, screen } from '@testing-library/react';

import {
  mockBuildURL,
  mockDefaultAPIBasePath,
  mockDefaultAPIServer,
  mockGetFromURLString,
  mockLogin,
  mockLoginStatus,
  mockLogout,
  mockPostToURLString,
  mockPrepareURLFromArgs,
  mockPutToURLString,
  mockRefreshToken,
} from '../../../../../Network/testing';

jest.mock('../../../../../Network/Login', () => {
  const originalModule = jest.requireActual('../../../../../Network/Login');

  return {
    __esModule: true,
    ...originalModule,
    loginStatus: mockLoginStatus,
    refreshToken: mockRefreshToken,
    login: mockLogin,
    logout: mockLogout,
  };
});

jest.mock('../../../../../Network/Network', () => {
  const originalModule = jest.requireActual('../../../../../Network/Network');

  return {
    __esModule: true,
    ...originalModule,
    defaultAPIServer: mockDefaultAPIServer,
    defaultAPIBasePath: mockDefaultAPIBasePath,
    buildURL: mockBuildURL,
    prepareURLFromArgs: mockPrepareURLFromArgs,
    getFromURLString: mockGetFromURLString,
    postToURLString: mockPostToURLString,
    putToURLString: mockPutToURLString,
  };
});

import { mockIntersectionObserver } from '../../../../../Hooks/testing';
import { setupTestCollectionOneFetch } from '../../../../../Model/testing';

import { ModelPickerList } from './ModelPickerList';
import {
  ContextChanger,
  TestCollectionOne,
  TestCollectionOneURL,
} from '../../../../../testHelpers';

jest.useFakeTimers();

const setupLogin = (loggedIn = true) => {
  const loginPromise = PromiseWithResolvers();
  mockLoginStatus.mockReturnValue(loginPromise);
  loginPromise.resolve({
    loggedIn,
    token: testToken,
  });
};

const testToken =
  'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJGREM4MTEzOCIsInVzZXIiOnsiaWQiOjEsIm5hbWUiOiJhZG1pbiIsImFkbWluIjpmYWxzZX0sImlhdCI6MTY2NTk2NTA5OX0.2vz14X7Tm-oFlyOa7dcAF-5y5ympi_UlWyJNxO4xyS4';

beforeEach(() => {
  window.IntersectionObserver = mockIntersectionObserver;
  mockBuildURL.mockReturnValue(TestCollectionOneURL);
});

afterEach(() => {
  delete window.IntersectionObserver;
  mockIntersectionObserver._clear();
});

describe('Test ModelPickerList', () => {
  it('component should match snapshot', async () => {
    const [getPromise, fetchBody] = setupTestCollectionOneFetch();

    setupLogin();

    const testCollection = new TestCollectionOne({});
    const { asFragment } = render(
      <ContextChanger initialLoggedIn={true}>
        <ModelPickerList initialCollection={testCollection} />
      </ContextChanger>,
    );

    await act(async () => {
      getPromise.resolve(fetchBody);
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the list as a popover dialog', async () => {
    const [getPromise, fetchBody] = setupTestCollectionOneFetch();

    setupLogin();

    const testCollection = new TestCollectionOne({});
    render(
      <ContextChanger initialLoggedIn={true}>
        <ModelPickerList initialCollection={testCollection} />
      </ContextChanger>,
    );

    const listDialog = screen.getByTestId('model-picker-list');
    expect(listDialog).toBeInTheDocument();
    expect(listDialog.tagName).toEqual('DIALOG');
    expect(listDialog).toHaveAttribute('id', 'model-picker-list');
    expect(listDialog).toHaveClass('model-picker-list');
    expect(listDialog).toHaveAttribute('popover', 'auto');

    const testList = screen.getByTestId('model-picker-list-list');
    expect(listDialog).toContainElement(testList);
    expect(testList).toBeEmptyDOMElement();

    await act(async () => {
      getPromise.resolve(fetchBody);
    });

    expect(testList).not.toBeEmptyDOMElement();
  });
});
