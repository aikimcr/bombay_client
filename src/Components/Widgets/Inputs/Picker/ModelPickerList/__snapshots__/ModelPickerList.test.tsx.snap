// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Test ModelPickerList component should match snapshot 1`] = `
<DocumentFragment>
  <dialog
    class="model-picker-list"
    data-testid="model-picker-list"
    id="model-picker-list"
    popover="auto"
  >
    <ul
      class="model-picker-list__list"
      data-testid="model-picker-list-list"
    >
      <li
        data-modelid="1"
      >
        fellow
      </li>
      <li
        data-modelid="2"
      >
        property
      </li>
      <li
        data-modelid="3"
      >
        yet
      </li>
      <li
        data-modelid="4"
      >
        article
      </li>
      <li
        data-modelid="5"
      >
        held
      </li>
      <li
        data-modelid="6"
      >
        transportation
      </li>
      <li
        data-modelid="7"
      >
        scene
      </li>
      <li
        data-modelid="8"
      >
        current
      </li>
      <li
        data-modelid="9"
      >
        life
      </li>
      <li
        data-modelid="10"
      >
        printed
      </li>
    </ul>
  </dialog>
  <button
    class="change-test-login"
    data-testid="change-test-login"
  >
    Change Login
  </button>
</DocumentFragment>
`;
